derive_miniconf/src/lib.rs

```diff
@@ -49,162 +37,157 @@ impl TypeDefinition {
         }
-    });
+    }
+}
 
-    let get_path_arms = fields.iter().map(|f| {
-        let match_name = &f.field.ident;
-        if f.deferred {
-            quote! {
-                stringify!(#match_name) => {
-                    self.#match_name.get_path(path_parts, value)
-                }
-            }
-        } else {
-            quote! {
-                stringify!(#match_name) => {
-                    if path_parts.peek().is_some() {
-                        return Err(miniconf::Error::PathTooLong);
-                    }
+fn next_path_arm((i, f): (usize, &StructField)) -> proc_macro2::TokenStream {
+    let field_type = &f.field.ty;
+    let field_name = &f.field.ident;
+    if f.deferred {
+        quote! {
+            #i => {
+                path.push_str(concat!(stringify!(#field_name), "/"))
+                    .map_err(|_| miniconf::IterError::PathLength)?;
 
-                    Ok(miniconf::serde_json_core::to_slice(&self.#match_name, value)?)
+                if <#field_type>::next_path(&mut state[1..], path)? {
+                    return Ok(true);
                 }
             }
         }
-    });
-
-    let next_path_arms = fields.iter().enumerate().map(|(i, f)| {
-        let field_type = &f.field.ty;
-        let field_name = &f.field.ident;
-        if f.deferred {
-            quote! {
-                #i => {
-                    let original_length = path.len();
-
-                    path.push_str(concat!(stringify!(#field_name), "/"))
-                        .map_err(|_| miniconf::IterError::PathLength)?;
-
-                    if <#field_type>::next_path(&mut state[1..], path)? {
-                        return Ok(true);
-                    }
-
-                    // Strip off the previously prepended index, since we completed that element and need
-                    // to instead check the next one.
-                    path.truncate(original_length);
-
-                    state[0] += 1;
-                    state[1..].fill(0);
-                }
-            }
-        } else {
-            quote! {
-                #i => {
-                    path.push_str(stringify!(#field_name))
-                        .map_err(|_| miniconf::IterError::PathLength)?;
-                    state[0] += 1;
-
-                    return Ok(true);
-                }
+    } else {
+        quote! {
+            #i => {
+                path.push_str(stringify!(#field_name))
+                    .map_err(|_| miniconf::IterError::PathLength)?;
+                state[0] += 1;
+
+                return Ok(true);
             }
         }
-    });
+    }
+}
 
-    let metadata_arms = fields.iter().enumerate().map(|(i, f)| {
-        let field_type = &f.field.ty;
-        let field_name = &f.field.ident;
-        if f.deferred {
-            quote! {
-                #i => {
-                    let mut meta = <#field_type>::metadata();
+fn metadata_arm((i, f): (usize, &StructField)) -> proc_macro2::TokenStream {
```

Comment:
Should we make these functions members of `StructField` instead?

Reply:
That would not help with your concern https://github.com/quartiq/miniconf/pull/116/files#r1015249137

Reply:
Scratch this. I was thinking `structfield` was a type we defined, but it's pulled in from `syn`

Reply:
But otherwise yes. Ich think that's nicer.